### PR TITLE
Cloudinary. Image upload/destroy for Vendor and Discount entities

### PR DESCRIPTION
**Pull request by VolodymyrBushko**

### CODE REVIEW

File: src/main/java/com/exadel/discountwebapp/cloud/CloudController.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
+package com.exadel.discountwebapp.cloud;
+
+import lombok.RequiredArgsConstructor;
+import org.springframework.web.bind.annotation.*;
+import org.springframework.web.multipart.MultipartFile;
+
+@RestController()
+@RequestMapping("/api/cloud")
+@RequiredArgsConstructor
+public class CloudController {
```
Comment (ahlukhouski2) [inline src/main/java/com/exadel/discountwebapp/cloud/CloudController.java]: Let's rename to something like ImageUploadController

File: src/main/java/com/exadel/discountwebapp/cloud/CloudController.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
+package com.exadel.discountwebapp.cloud;
+
+import lombok.RequiredArgsConstructor;
+import org.springframework.web.bind.annotation.*;
+import org.springframework.web.multipart.MultipartFile;
+
+@RestController()
+@RequestMapping("/api/cloud")
```
Comment (ahlukhouski2) [inline src/main/java/com/exadel/discountwebapp/cloud/CloudController.java]: let's rename to something like "/api/image-upload"

File: src/main/java/com/exadel/discountwebapp/cloud/CloudController.java [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+    private final ImageCloudService imageCloudService;
+
+    @PostMapping("/upload/image")
+    public String uploadImage(@RequestParam("image") MultipartFile image) {
+        return imageCloudService.upload(image);
+    }
+
+    @DeleteMapping("/destroy/image")
+    public boolean destroyImage(@RequestBody String url) {
```
Comment (ahlukhouski2) [inline src/main/java/com/exadel/discountwebapp/cloud/CloudController.java:20]: the image should be deleted by its id, not by url
we can pass this id in request url

Comment (ahlukhouski2) [inline src/main/java/com/exadel/discountwebapp/cloud/CloudController.java]: let's rename the method to deleteImage

File: src/main/java/com/exadel/discountwebapp/cloud/ImageCloudService.java [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+import org.springframework.web.multipart.MultipartFile;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+@Slf4j
+@Service
+public class ImageCloudService {
```
Comment (ahlukhouski2) [inline src/main/java/com/exadel/discountwebapp/cloud/ImageCloudService.java]: let's rename to ImageUploadService

File: src/main/java/com/exadel/discountwebapp/cloud/CloudController.java [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+
+@RestController()
+@RequestMapping("/api/cloud")
+@RequiredArgsConstructor
+public class CloudController {
+
+    private final ImageCloudService imageCloudService;
+
+    @PostMapping("/upload/image")
+    public String uploadImage(@RequestParam("image") MultipartFile image) {
```
Comment (ahlukhouski2) [inline src/main/java/com/exadel/discountwebapp/cloud/CloudController.java]: it's better return the response body with the image url and some params like extension, size

File: src/main/java/com/exadel/discountwebapp/discount/service/DiscountService.java [hunk 6]
```diff
@@ -60,16 +62,29 @@ public DiscountResponseVO create(DiscountRequestVO request) {
...
     public DiscountResponseVO update(Long id, DiscountRequestVO request) {
-        Discount discount = discountRepository.findById(id)
-                .orElseThrow(() -> new EntityNotFoundException(Discount.class, "id", id));
+        Discount discount = getDiscountById(id);
+        String imageUrl = discount.getImageUrl();
+
         discountMapper.updateEntity(request, discount);
         discountRepository.save(discount);
+
+        if (!imageUrl.equals(request.getImageUrl())) {
```
Comment (ahlukhouski2) [inline src/main/java/com/exadel/discountwebapp/discount/service/DiscountService.java:71]: it seems like now we should store discount image id, not url

File: src/main/java/com/exadel/discountwebapp/cloud/ImageCloudValidator.java [hunk 7]
```diff
@@ -0,0 +1,29 @@
+package com.exadel.discountwebapp.cloud;
+
+import com.exadel.discountwebapp.exception.exception.fileupload.FileEmptyException;
+import com.exadel.discountwebapp.exception.exception.fileupload.FileOverSizeException;
+import org.springframework.stereotype.Component;
+import org.springframework.web.multipart.MultipartFile;
+
+@Component
+public class ImageCloudValidator {
```
Comment (ahlukhouski2) [inline src/main/java/com/exadel/discountwebapp/cloud/ImageCloudValidator.java]: ImageUploadValidator

File: src/main/java/com/exadel/discountwebapp/vendor/service/VendorService.java [hunk 8]
```diff
@@ -59,18 +61,28 @@ public VendorResponseVO create(VendorRequestVO request) {
...
+
         if (!vendor.getEmail().equals(request.getEmail())) {
             vendorEmailValidator.validate(request);
         }
+
         vendorMapper.update(request, vendor);
-        var updatedVendor = vendorRepository.save(vendor);
+        Vendor updatedVendor = vendorRepository.save(vendor);
+
+        if (!imageUrl.equals(request.getImageUrl())) {
```
Comment (ahlukhouski2) [inline src/main/java/com/exadel/discountwebapp/vendor/service/VendorService.java:74]: imageUrl can be null

File: src/main/java/com/exadel/discountwebapp/discount/service/DiscountService.java (hunk 6 above)
Comment (ahlukhouski2) [inline src/main/java/com/exadel/discountwebapp/discount/service/DiscountService.java]: imageUrl can be null

File: src/main/java/com/exadel/discountwebapp/cloud/CloudController.java (hunk 1 above)
Comment (VolodymyrBushko) [reply src/main/java/com/exadel/discountwebapp/cloud/CloudController.java]: done

File: src/main/java/com/exadel/discountwebapp/cloud/CloudController.java (hunk 2 above)
Comment (VolodymyrBushko) [reply src/main/java/com/exadel/discountwebapp/cloud/CloudController.java]: done

File: src/main/java/com/exadel/discountwebapp/cloud/CloudController.java (hunk 3 above)
Comment (VolodymyrBushko) [reply src/main/java/com/exadel/discountwebapp/cloud/CloudController.java]: done

File: src/main/java/com/exadel/discountwebapp/cloud/ImageCloudService.java (hunk 4 above)
Comment (VolodymyrBushko) [reply src/main/java/com/exadel/discountwebapp/cloud/ImageCloudService.java]: done

File: src/main/java/com/exadel/discountwebapp/cloud/CloudController.java (hunk 5 above)
Comment (VolodymyrBushko) [reply src/main/java/com/exadel/discountwebapp/cloud/CloudController.java]: done

File: src/main/java/com/exadel/discountwebapp/cloud/ImageCloudValidator.java (hunk 7 above)
Comment (VolodymyrBushko) [reply src/main/java/com/exadel/discountwebapp/cloud/ImageCloudValidator.java]: done

File: src/main/java/com/exadel/discountwebapp/discount/service/DiscountService.java (hunk 6 above)
Comment (VolodymyrBushko) [reply src/main/java/com/exadel/discountwebapp/discount/service/DiscountService.java]: I added condition. The vendor has a non-null constraint, but Discount doesn't have

File: src/test/java/com/exadel/discountwebapp/vendor/service/VendorServiceIntegrationTest.java [hunk 9]
```diff
@@ -2,15 +2,18 @@
...
 import com.exadel.discountwebapp.exception.exception.client.IncorrectFilterInputException;
+import com.exadel.discountwebapp.fileupload.image.ImageUploadService;
 import com.exadel.discountwebapp.user.repository.UserRepository;
 import com.exadel.discountwebapp.vendor.entity.Vendor;
 import com.exadel.discountwebapp.vendor.repository.VendorRepository;
 import com.exadel.discountwebapp.vendor.vo.VendorRequestVO;
 import com.exadel.discountwebapp.vendor.vo.VendorResponseVO;
 import com.google.common.collect.Lists;
 import org.junit.jupiter.api.Test;
+import org.mockito.Mockito;
```
Comment (ahlukhouski2) [inline src/test/java/com/exadel/discountwebapp/vendor/service/VendorServiceIntegrationTest.java]: redundant import

File: src/test/java/com/exadel/discountwebapp/discount/controller/DiscountControllerIntegrationTest.java [hunk 10]
```diff
@@ -12,9 +13,11 @@
 import org.junit.jupiter.api.Assertions;
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.extension.RegisterExtension;
+import org.mockito.Mockito;
```
Comment (ahlukhouski2) [inline src/test/java/com/exadel/discountwebapp/discount/controller/DiscountControllerIntegrationTest.java]: redundant import

File: src/test/java/com/exadel/discountwebapp/discount/service/DiscountServiceIntegrationTest.java [hunk 11]
```diff
@@ -16,8 +17,10 @@
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.extension.RegisterExtension;
+import org.mockito.Mockito;
```
Comment (ahlukhouski2) [inline src/test/java/com/exadel/discountwebapp/discount/service/DiscountServiceIntegrationTest.java]: redundant import

Comment (VolodymyrBushko) [reply src/test/java/com/exadel/discountwebapp/discount/service/DiscountServiceIntegrationTest.java]: done

File: src/test/java/com/exadel/discountwebapp/discount/controller/DiscountControllerIntegrationTest.java (hunk 10 above)
Comment (VolodymyrBushko) [reply src/test/java/com/exadel/discountwebapp/discount/controller/DiscountControllerIntegrationTest.java]: done

File: src/test/java/com/exadel/discountwebapp/vendor/service/VendorServiceIntegrationTest.java (hunk 9 above)
Comment (VolodymyrBushko) [reply src/test/java/com/exadel/discountwebapp/vendor/service/VendorServiceIntegrationTest.java]: done

**Review by ahlukhouski2 — APPROVED** [review]

**Review by Bozhokmaria — APPROVED** [review]

**Pull request merged**